runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/FnApiControlClientPoolService.java

```diff
@@ -105,12 +105,38 @@ public static FnApiControlClientPoolService offeringClientsToPool(
     return newClient.asResponseObserver();
   }
 
+  @Override
+  public void getProcessBundleDescriptor(
+      BeamFnApi.GetProcessBundleDescriptorRequest request,
+      StreamObserver<BeamFnApi.ProcessBundleDescriptor> responseObserver) {
+    final String workerId = headerAccessor.getSdkWorkerId();
+    if (Strings.isNullOrEmpty(workerId)) {
+      // TODO(BEAM-4149): Enforce proper worker id.
+      LOG.warn("No worker_id header provided in getProcessBundleDescriptor request");
+    }
+
+    LOG.info("getProcessBundleDescriptor request with id {}", workerId);
+
+    FnApiControlClient client;
+    synchronized (lock) {
+      client = vendedClients.get(workerId);
+    }
+    LOG.warn("getProcessBundleDescriptor client {}", client);
+    LOG.warn("getProcessBundleDescriptor id {}", request.getProcessBundleDescriptorId());
+    LOG.warn(
+        "getProcessBundleDescriptor result {}",
+        client.getProcessBundleDescriptor(request.getProcessBundleDescriptorId()));
+    responseObserver.onNext(
+        client.getProcessBundleDescriptor(request.getProcessBundleDescriptorId()));
```

Comment:
Please return the NOT_FOUND status error if the process bundle descriptor is null.